src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,177 @@
+        public virtual void Initialize()
+        {
+            _firmata = new FirmataDevice();
+            _firmata.Open(_serialPortStream);
+            _firmata.OnError += FirmataOnError;
+            var protocolVersion = _firmata.QueryFirmataVersion();
+            if (protocolVersion != _firmata.QuerySupportedFirmataVersion())
+            {
+                throw new NotSupportedException($"Firmata version on board is {protocolVersion}. Expected {_firmata.QuerySupportedFirmataVersion()}. They must be equal.");
+            }
+
+            Log($"Firmata version on board is {protocolVersion}.");
+
+            _firmwareVersion = _firmata.QueryFirmwareVersion(out _firmwareName);
+
+            Log($"Firmware version on board is {_firmwareVersion}");
+
+            _firmata.QueryCapabilities();
+
+            _supportedPinConfigurations = _firmata.PinConfigurations; // Clone reference
+
+            Log("Device capabilities: ");
+            foreach (var pin in _supportedPinConfigurations)
+            {
+                Log(pin.ToString());
+            }
+
+            // _firmata.SetSamplingInterval(TimeSpan.FromMilliseconds(100));
+            _firmata.EnableDigitalReporting();
+        }
+
+        public Version FirmwareVersion
+        {
+            get
+            {
+                return _firmwareVersion;
+            }
+        }
+
+        public string FirmwareName
+        {
+            get
+            {
+                return _firmwareName;
+            }
+        }
+
+        internal FirmataDevice Firmata
+        {
+            get
+            {
+                return _firmata;
+            }
+        }
+
+        internal List<SupportedPinConfiguration> SupportedPinConfigurations
+        {
+            get
+            {
+                return _supportedPinConfigurations;
+            }
+        }
+
+        internal void Log(string message)
+        {
+            LogMessages?.Invoke(message, null);
+        }
+
+        private void FirmataOnError(string message, Exception innerException)
+        {
+            LogMessages?.Invoke(message, innerException);
+        }
+
+        public GpioController CreateGpioController(PinNumberingScheme pinNumberingScheme)
+        {
+            return new GpioController(pinNumberingScheme, new ArduinoGpioControllerDriver(this, _supportedPinConfigurations));
+        }
+
+        public I2cDevice CreateI2cDevice(I2cConnectionSettings connectionSettings)
+        {
+            return new ArduinoI2cDevice(this, connectionSettings);
+        }
+
+        /// <summary>
+        /// Firmata has no support for SPI, even though the Arduino basically has an SPI interface.
+        /// This therefore returns a Software SPI device for the default Arduino SPI port on pins 11, 12 and 13.
+        /// </summary>
+        /// <param name="settings">Spi Connection settings</param>
+        /// <returns></returns>
+        public SpiDevice CreateSpiDevice(SpiConnectionSettings settings)
```

Comment:
There's an existing protocol proposal for SPI over Firmata, but ASFAIK no implementation yet. It would probably not be to hard to add that implementation to ConfigurableFirmata. 
I thought SoftwareSPI was better than nothing, but I did not test this in any way yet.

Reply:
What is the license associated to this ino file? My guess is that regarding the type of license from the original Firmata repo, doesn't seems very compatible with this repo.

Reply:
The license for ConfigurableFirmata is LGPL. So coding against it as a library is fine. [Firmata-Builder](https://github.com/firmata/firmata-builder) (the generator that created this main application file) has a MIT license. So re-distribution (even with changes) inside this repo is OK.

Reply:
I just tested software SPI over firmata with an MCP3008. It does not really work 😞 . It manages to run the clock at ~1.25kHz, with a big skew, while it normally runs at 1Mhz. 
I can try whether I get it to run somehow, but for now, I think we should declare this an unsupported feature for Arduino.

Reply:
I was unfortunately expecting it not to work :-( Have you try the native implementation of SPI in Firmata or just the software SPI?

Reply:
The bad part: The problem is that Firmata has no SPI implementation at all. So far there's only a protocol proposal, but no implementation for it (except maybe a prototype for StandardFirmata). 
The good part: I've quickly written an implementation for that proposal yesterday and that _does_ work. The maintainer of the firmata repos is accepting PRs, even if he's not actively working on it right now. 
In either case: This will come down to query whether the Arduino has SPI support and use it in that case or throw if not. That's what I should do anyway, because even if an implementation exists, it could be disabled.

Reply:
That's great news and that would make a huge difference. Native SPI is used a lot and very much needed for most of the bindings we have. Great 😊 Looking forward to try it 😊